models/image.py

```diff
@@ -111,65 +111,46 @@ def format(self, value):
     )
 
-    ref_image_index = sa.Column(
-        sa.Integer,
+    downstream_images = orm.relationship(
+        'Image',
+        secondary=image_upstreams_association_table,
+        primaryjoin='images.c.id == image_upstreams_association.c.upstream_id',
+        secondaryjoin='images.c.id == image_upstreams_association.c.downstream_id',
+        overlaps="upstream_images",
+        cascade='save-update, merge, refresh-expire, expunge',
+        passive_deletes=True,
+        order_by='images.c.mjd',  # in chronological order of exposure start time
+        doc='Combined Images (like coadds or a subtractions) that use this image in their production. '
+    )
+
+    ref_image_id = sa.Column(
+        sa.ForeignKey('images.id', ondelete="SET NULL", name='images_ref_image_id_fkey'),
         nullable=True,
+        index=True,
         doc=(
-            "Index of the reference image used to produce this image, in the upstream_images list. "
+            "ID of the reference image used to produce this image, in the upstream_images list. "
         )
     )
 
-    @property
-    def ref_image(self):
-        if self.ref_image_index is None:
-            return None
-        if len(self.upstream_images) <= self.ref_image_index:
-            raise RuntimeError(f'Index {self.ref_image_index} is out of range for upstream images!')
-        return self.upstream_images[self.ref_image_index]
-
-    @ref_image.setter
-    def ref_image(self, value):
-        if value is None:
-            self.ref_image_index = None
-        else:
-            if not isinstance(value, Image):
-                raise ValueError(f"ref_image must be an Image object. Got {type(value)} instead.")
-            if value not in self.upstream_images:
-                raise ValueError(f"ref_image must be in the upstream_images list. Got {value} instead.")
-
-            # make sure the upstream_images list is sorted by mjd:
-            self.upstream_images.sort(key=lambda x: x.mjd)
-            self.ref_image_index = self.upstream_images.index(value)
-
-    new_image_index = sa.Column(
-        sa.Integer,
-        nullable=True,
+    ref_image = orm.relationship(
+        'Image',
+        primaryjoin='Image.ref_image_id == Image.id',
+        remote_side='Image.id',
+        cascade='save-update, merge, refresh-expire, expunge',
+        uselist=False,
+        lazy='selectin',
         doc=(
-            "Index of the new image used to produce a difference image, in the upstream_images list. "
+            "Reference image used to produce this image, in the upstream_images list. "
         )
     )
 
     @property
     def new_image(self):
-        if self.new_image_index is None:
+        """Get the image that is NOT the reference image. This only works on subtractions (with ref+new upstreams)"""
+        image = [im for im in self.upstream_images if im.id != self.ref_image_id]
+        if len(image) == 0 or len(image) > 1:
```

Comment:
If len(image) > 1, we should raise an exception here, rather than return None ; new_image() was called on a coadd rather than a subtraction, so new_image is meaningless.

Reply:
The problem is that `new_image` is not a function, it is a property. I think calling properties that routinely raise errors is going to be a headache (consider listing all properties of an  object). When it returns none you just know it is not relevant.